cpp/ql/test/query-tests/Security/CWE/CWE-570/test.cpp

```diff
@@ -233,3 +233,54 @@ void test_operator_new_without_exception_spec() {
   int* p = new(42, std::nothrow) int; // GOOD
   if(p == nullptr) {}
 }
+
+namespace std {
+  void *memset(void *s, int c, size_t n);
+}
+
+// from the qhelp:
+namespace qhelp {
+  // BAD: the allocation will throw an unhandled exception
+  // instead of returning a null pointer.
+  void bad1(std::size_t length) noexcept {
+    int* dest = new int[length];
+    if(!dest) {
+      return;
+    }
+    std::memset(dest, 0, length);
+    // ...
+  }
+
+  // BAD: the allocation won't throw an exception, but
+  // instead return a null pointer. [NOT DETECTED]
+  void bad2(std::size_t length) noexcept {
+    try {
+      int* dest = new(std::nothrow) int[length];
+      std::memset(dest, 0, length);
```

Comment:
As far as I can tell there is no guarantee that `std::memcpy` _doesn't_ throw in the language standard, but most implementations won't.  The one I'm looking appears to get marked `nothrow` if it's compiled as C++, but not if it's compiled as C, and it appears to get C linkage and not `nothrow` in the actual database.  It also has a body that just calls a builtin function.

So we could perhaps address this by assuming that a C linkage function doesn't throw ... but I don't think it's actually strictly true that a C linkage function can't throw.  Or that builtin function don't throw, but again I'm not convinced that's true.  Which leaves us with modelling ... we could model functions that don't throw, but again I don't know if it's strictly true that all versions of `memset` don't throw, and it seems a bit heavy handed to model it given that `nothrow` exists.

Reply:
`std::memcpy` is from the C standard library, so it cannot throw. But you're right that it's not marked as `noexcept` in the standard.

Reply:
And this is probably the relevant part of the standard:

<img width="808" alt="Screenshot 2024-07-10 at 14 07 22" src="https://github.com/user-attachments/assets/bff2f867-34e6-46d1-855a-a74b64715045">

So since `memcpy` cannot throw in C, and the contents of `cstring` and `string.h` is identical, my reading is that `memcpy` also isn't allowed to throw 🤔

Reply:
You make a good case for modelling.  I'll create an issue for this.

Reply:
Issue created.  Thanks for helping me get to the bottom of this!